README.md
Suplemon :lemon:
========

[![Build Status](https://travis-ci.org/richrd/suplemon.svg?branch=master)](https://travis-ci.org/richrd/suplemon) [![Join the chat at https://webchat.freenode.net/?channels=%23suplemon](https://img.shields.io/badge/chat-on%20freenode%20%23suplemon-blue.svg)](https://webchat.freenode.net/?channels=%23suplemon)

          ___________   _________  ___     ______________________________   ___
         /  _____/  /  /  /  _   \/  /\   /  ______/        /  ___   /   | /  /\
        /  /____/  /  /  /  /_/  /  / /  /  /_____/  /  /  /  /  /  /    |/  / /
       /____   /  /  /  /  _____/  / /  /  ______/  /  /  /  /  /  /  /|    / /
      _____/  /  /__/  /  /\___/  /____/  /_____/  /  /  /  /__/  /  / |   / /
     /_______/\_______/__/ /  /_______/________/__/__/__/________/__/ /|__/ /
     \_______\ \______\__\/   \_______\________\__\__\__\________\__\/ \__\/

              Remedying the pain of command line editing since 2014


Suplemon is a modern, powerful and intuitive console text editor with multi cursor support.
Suplemon replicates Sublime Text style functionality in the terminal with the ease of use of Nano.
http://github.com/richrd/suplemon

![Suplemon in action](http://bittemple.org/sites/suplemon.com/assets/suplemon-v0.1.51-multi-cursor-editing.gif)

## Features
 * Proper multi cursor editing, as in Sublime Text
 * Syntax highlighting with Text Mate themes
 * Autocomplete (based on words in the files that are open)
 * Easy Undo/Redo (Ctrl + Z, Ctrl + Y)
 * Copy & Paste, with multi line support (and native clipboard support on X11 / Unix and Mac OS)
 * Multiple files in tabs
 * Powerful Go To feature for jumping to files and lines
 * Find, Find next and Find all (Ctrl + F, Ctrl + D, Ctrl + A)
 * Custom keyboard shortcuts (and easy-to-use defaults)
 * Mouse support
 * Restores cursor and scroll positions when reopenning files
 * Extensions (easy to write your own)
 * Lots more...


## Caveats
 * Currently no built in selections (regions). To copy part of a line select it with your mouse and use <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>C</kbd>

## Try it!

You can just clone the repo, and try Suplemon, or also install it system wide.
To run from source you need to install the python `wcwidth` package.

    pip3 install wcwidth
    git clone https://github.com/richrd/suplemon.git
    cd suplemon
    python3 suplemon.py

### Installation

Install the latest version from PIP:

    sudo pip3 install suplemon

To install Suplemon from the repo run the setup script:

    sudo python3 setup.py install

### Usage

    suplemon # New file in the current directory
    suplemon [filename]... # Open one or more files
    suplemon [filename:row:col]... # Open one or more files at a specific row or column (optional)


### Notes
 - **Must use Python 3.3 or higher for proper character encoding support.**
 - **Python2.7 (and maybe lower) versions work, but aren't officially supported (some special characters won't work etc).**
 - *The master branch is considered stable.*
 - *Tested on Unix.*

Dev Branch Status: [![Build Status](https://travis-ci.org/richrd/suplemon.svg?branch=dev)](https://travis-ci.org/richrd/suplemon)

No dependencies outside the Python Standard Library required.

### Optional dependencies

 * Pygments
 > For support for syntax highlighting over 300 languages.

 * Flake8
 > For showing linting for Python files.

 * xsel or xclip
 > For system clipboard support on X Window (Linux).

 * pbcopy / pbpaste
 > For system clipboard support on Mac OS.

 See [docs/optional-dependencies.md][] for installation instructions.

 [docs/optional-dependencies.md]: docs/optional-dependencies.md

## Description
Suplemon is an intuitive command line text editor. It supports multiple cursors out of the box.
It is as easy as nano, and has much of the power of Sublime Text. It also supports extensions
to allow all kinds of customizations. To get more help hit ```Ctrl + H``` in the editor.
Suplemon is licensed under the MIT license.

## Configuration

### Main Config
The suplemon config file is stored at ```~/.config/suplemon/suplemon-config.json```.

The best way to edit it is to run the ```config``` command (Run commands via ```Ctrl+E```).
That way Suplemon will automatically reload the configuration when you save the file.
To view the default configuration and see what options are available run ```config defaults``` via ```Ctrl+E```.


### Keymap Config

Below are the default key mappings used in suplemon. They can be edited by running the ```keymap``` command.
To view the default keymap file run ```keymap default```

 * <kbd>Ctrl</kbd> + <kbd>Q</kbd>
   > Exit

 * <kbd>Ctrl</kbd> + <kbd>W</kbd>
   > Close file or tab

 * <kbd>Ctrl</kbd> + <kbd>C</kbd>
   > Copy line(s) to buffer

 * <kbd>Ctrl</kbd> + <kbd>X</kbd>
   > Cut line(s) to buffer

 * <kbd>Ctrl</kbd> + <kbd>V</kbd>
   > Insert buffer

 * <kbd>Ctrl</kbd> + <kbd>K</kbd>
   > Duplicate line

 * <kbd>Ctrl</kbd> + <kbd>G</kbd>
   > Go to line number or file (type the beginning of a filename to switch to it).
   > You can also use 'filena:42' to go to line 42 in filename.py etc.

 * <kbd>Ctrl</kbd> + <kbd>F</kbd>
   > Search for a string or regular expression (configurable)

 * <kbd>Ctrl</kbd> + <kbd>D</kbd>
   > Search for next occurrence or find the word the cursor is on. Adds a new cursor at each new occurrence.

 * <kbd>Ctrl</kbd> + <kbd>T</kbd>
   > Trim whitespace

 * <kbd>Alt</kbd> + <kbd>Arrow Key</kbd>
   > Add new cursor in arrow direction

 * <kbd>Ctrl</kbd> + <kbd>Left / Right</kbd>
   > Jump to previous or next word or line

 * <kbd>ESC</kbd>
   > Revert to a single cursor / Cancel input prompt

 * <kbd>Alt</kbd> + <kbd>Page Up</kbd>
   > Move line(s) up

 * <kbd>Alt</kbd> + <kbd>Page Down</kbd>
   > Move line(s) down

 * <kbd>Ctrl</kbd> + <kbd>S</kbd>
   > Save current file

 * <kbd>F1</kbd>
   > Save file with new name

 * <kbd>F2</kbd>
   > Reload current file

 * <kbd>Ctrl</kbd> + <kbd>O</kbd>
   > Open file

 * <kbd>Ctrl</kbd> + <kbd>W</kbd>
   > Close file

 * <kbd>Ctrl</kbd> + <kbd>Page Up</kbd>
   > Switch to next file

 * <kbd>Ctrl</kbd> + <kbd>Page Down</kbd>
   > Switch to previous file

 * <kbd>Ctrl</kbd> + <kbd>E</kbd>
   > Run a command.

 * <kbd>Ctrl</kbd> + <kbd>Z</kbd> and <kbd>F5</kbd>
   > Undo

 * <kbd>Ctrl</kbd> + <kbd>Y</kbd> and <kbd>F6</kbd>
   > Redo

 * <kbd>F7</kbd>
   > Toggle visible whitespace

 * <kbd>F8</kbd>
   > Toggle mouse mode

 * <kbd>F9</kbd>
   > Toggle line numbers

 * <kbd>F11</kbd>
   > Toggle full screen

## Mouse shortcuts

 * Left Click
   > Set cursor at mouse position. Reverts to a single cursor.

 * Right Click
   > Add a cursor at mouse position.

 * Scroll Wheel Up / Down
   > Scroll up & down.

## Commands

Suplemon has various add-ons that implement extra features.
The commands can be run with <kbd>Ctrl</kbd> + <kbd>E</kbd> and the prompt has autocomplete to make running them faster.
The available commands and their descriptions are:

 * autocomplete

    A simple autocompletion module.

    This adds autocomplete support for the tab key. It uses a word
    list scanned from all open files for completions. By default it suggests
    the shortest possible match. If there are no matches, the tab action is
    run normally.

 * autodocstring

    Simple module for adding docstring placeholders.

    This module is intended to generate docstrings for Python functions.
    It adds placeholders for descriptions, arguments and return data.
    Function arguments are crudely parsed from the function definition
    and return statements are scanned from the function body.

 * bulk_delete

    Bulk delete lines and characters.
    Asks what direction to delete in by default.

    Add 'up' to delete lines above highest cursor.
    Add 'down' to delete lines below lowest cursor.
    Add 'left' to delete characters to the left of all cursors.
    Add 'right' to delete characters to the right of all cursors.

 * comment

    Toggle line commenting based on current file syntax.

 * config

    Shortcut for openning the config files.

 * crypt

    Encrypt or decrypt the current buffer. Lets you provide a passphrase and optional salt for encryption.
    Uses AES for encryption and scrypt for key generation.

 * diff

    View a diff of the current file compared to it's on disk version.

 * eval

    Evaluate a python expression and show the result in the status bar.

    If no expression is provided the current line(s) are evaluated and
    replaced with the evaluation result.

 * keymap

    Shortcut to openning the keymap config file.

 * linter

    Linter for suplemon.

 * lower

    Transform current lines to lower case.

 * lstrip

    Trim whitespace from beginning of current lines.

 * paste

    Toggle paste mode (helpful when pasting over SSH if auto indent is enabled)

 * reload

    Reload all add-on modules.

 * replace_all

    Replace all occurrences in all files of given text with given replacement.

 * reverse

    Reverse text on current line(s).

 * rstrip

    Trim whitespace from the end of lines.

 * save

    Save the current file.

 * save_all

    Save all currently open files. Asks for confirmation.

 * sort_lines

    Sort current lines.

    Sorts alphabetically by default.
    Add 'length' to sort by length.
    Add 'reverse' to reverse the sorting.

 * strip

    Trim whitespace from start and end of lines.

 * tabstospaces

    Convert tab characters to spaces in the entire file.

 * toggle_whitespace

    Toggle visually showing whitespace.

 * upper

    Transform current lines to upper case.


## Support

If you experience problems, please submit a new issue.
If you have a question, need help, or just want to chat head over to the IRC channel #suplemon @ Freenode.
I'll be happy to chat with you, see you there!


## Development

If you are interested in contributing to Suplemon, development dependencies can be installed via:

    # For OS cleanliness, we recommend using `virtualenv` to prevent global contamination
    pip install -r requirements-dev.txt

After those are installed, tests can be run via:

    ./test.sh

PRs are very welcome and appreciated.
When making PRs make sure to set the target branch to `dev`. I only push to master when releasing new versions.


## Rationale
For many the command line is a different environment for text editing.
Most coders are familiar with GUI text editors and for many vi and emacs
have a too steep learning curve. For them (like for me) nano was the weapon of
choice. But nano feels clunky and it has its limitations. That's why
I wrote my own editor with built in multi cursor support to fix the situation.
Another reason is that developing Suplemon is simply fun to do.


CHANGELOG.md
Change Log
==========


## [v0.2.1](https://github.com/richrd/suplemon/tree/0.2.1) (2019-08-29) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.2.0...0.2.1)

**Fixed bugs:**

- Fix a bug introduced by 0.2.0, where key bindings weren't set up if the user key config file was missing.

## [v0.2.0](https://github.com/richrd/suplemon/tree/0.2.0) (2019-06-16) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.1.65...0.2.0)

**Fixed bugs:**

- Fix not using the delta argument in the cursor move_up method.
- Fix issue where mouse events in prompts could crash suplemon #247
- Fix not being able to override default keys with user key bindings

**Implemented enhancements:**

- Allow help to be toggled with the help shortcut. Credit @caph1993
- Allow opening files at specific row and column from command line.


## [v0.1.65](https://github.com/richrd/suplemon/tree/0.1.65) (2019-03-11) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.1.64...0.1.65)

**Fixed bugs:**

- Merge all the default config into user config. Credit @Consolatis
- Fix diff command highlighting
- Add shift+pageup and shift+pagedown bindings
- Reuse existing windows on resize. Credit @Consolatis
- Replace link to gitter chat with freenode webchat. Credit @Consolatis
- Merge default module configs into user config. Credit @Consolatis
- Use daemon mode for lint thread to fix shutdown delay. Credit @Consolatis
- Simple xterm-256color imitation for curses. Credit @abl

**Implemented enhancements:**

- Allow line number padding using spaces instead of zeros. Credit @Consolatis
- Highlight current line(s) Credit @Consolatis
- Add crypt module for encrypting buffers with a password


## [v0.1.64](https://github.com/richrd/suplemon/tree/0.1.64) (2017-12-17) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.1.63...0.1.64)

**Implemented enhancements:**

- Add bulk_delete and sort_lines commands.
- Lots of code style fixes and improvements. Credit @Gnewbee
- Add xclip support for system clipboard. Credit @LChris314
- Added command docs to readme and help.


## [v0.1.63](https://github.com/richrd/suplemon/tree/0.1.63) (2017-10-05) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.1.62...0.1.63)

**Implemented enhancements:**

- Add autocomplete to run command prompt (fixes #171)
- Increase battery status polling time to 60 sec (previously 10 sec)
- Change the top bar suplemon icon to a fancy unicode lemon.
- Add paste mode for better pasting over SSH (disables auto indentation)

**Fixed bugs:**

- Keep top bar statuses of modules in alphabetical order based on module name. (fixes #57)
- Prevent restoring file state if file has changed since last time (fixes #198)


## [v0.1.62](https://github.com/richrd/suplemon/tree/0.1.62) (2017-09-25) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.1.61...0.1.62)

**Fixed bugs:**

- Fixed and re-enabled fancy unicode symbols.
- Fixed typos in default configuration. Credit @1fabunicorn (#192)
- Fixed error message when loading valid but empty config file (Fixes #196).

**Implemented enhancements:**

- Add ctrl+t shortcut for trimming whitespace


## [v0.1.61](https://github.com/richrd/suplemon/tree/0.1.61) (2017-05-29) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.1.60...0.1.61)

**Fixed bugs:**

- Disable fancy unicode symbols by default. Caused problems on some terminals.


## [v0.1.60](https://github.com/richrd/suplemon/tree/0.1.60) (2017-03-23) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.1.59...0.1.60)

**Implemented enhancements:**

- Add support for the MacOS native pasteboard via pbcopy/pbpaste. Credit @abl
- Added shift+tab for going backwards when autocompleting files.
- Added F keys with modifiers and fixed some ctrl+shift keybindings.

**Fixed bugs:**

- Broader error handling in hostname module.
- Don't print log message when opening file that doesn't exist.


## [v0.1.59](https://github.com/richrd/suplemon/tree/0.1.59) (2017-02-16) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.1.58...0.1.59)

**Implemented enhancements:**

- Added pygments as a dependency. 


**Fixed bugs:**

- Added error handling to hostname module should it fail.


## [v0.1.58](https://github.com/richrd/suplemon/tree/0.1.58) (2016-12-01) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.1.57...0.1.58)

**Fixed bugs:**

- Fixed tests by using newer flake8.
- Treat .ts files the same as .js so (un)commenting .ts files works.


## [v0.1.57](https://github.com/richrd/suplemon/tree/0.1.57) (2016-11-01) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.1.56...0.1.57)

**Implemented enhancements:**
- Show all special unicode whitespace characters when "show_white_space" is set to true. This helps detecting unwanted characters in files.
- Now the mouse wheel works the same way in normal mode and mouse mode.

**Fixed bugs:**

- Fixed weird crash when pasting lots of text into prompts (like the find prompt etc)
- Fixed false matches in diff highlighting.
- Fixed inability to open empty files.
- Fixed adding cursors via mouse click when the view is horizontally scrolled.
- Fixed inputting various special characters that were ignored in some cases on Python3.


## [v0.1.56](https://github.com/richrd/suplemon/tree/0.1.56) (2016-08-01) compared to previous master branch.
[Full Changelog](https://github.com/richrd/suplemon/compare/0.1.55...0.1.56)

**Implemented enhancements:**

- New feature: Ability to use hard tabs instead of spaces via the boolean option 'hard_tabs'.

ui.py
# -*- encoding: utf-8
"""
Curses user interface.
"""

import os

from helpers import *


class InputEvent:
    """Represents a keyboard or mouse event."""
    def __init__(self):
        self.type = None  # 'key' or 'mouse'
        self.key_name = None
        self.key_code = None
        self.mouse_code = None
        self.mouse_pos = (0, 0)

    def parse_key_code(self, code):
        """Parse a key namoe or code from curses."""
        self.type = "key"
        self.key_code = code
        self.key_name = self._key_name(code)

    def set_key_name(self, name):
        """Manually set the event key name."""
        self.type = "key"
        self.key_name = name

    def parse_mouse_state(self, state):
        """Parse curses mouse events."""
        self.type = "mouse"
        self.mouse_code = state[4]
        self.mouse_pos = (state[1], state[2])

    def _key_name(self, key):
        """Return the curses key name for keys received from get_wch (and getch)."""
        # Handle multibyte get_wch input in Python 3.3
        if isinstance(key, str):
            return str(curses.keyname(ord(key)).decode("utf-8"))
        # Fallback to try and handle Python < 3.3
        if isinstance(key, int):  # getch fallback
            try:  # Try to convert to a curses key name
                return str(curses.keyname(key).decode("utf-8"))
            except:  # Otherwise try to convert to a character
                try:
                    return chr(key)
                except:
                    return False
        return key

    def __str__(self):
        parts = [
            str(self.type),
            str(self.key_name),
            str(self.key_code),
            str(self.mouse_code),
            str(self.mouse_pos)
        ]
        return " ".join(parts)


class UI:
    def __init__(self, app):
        self.app = app
        self.warned_old_curses = 0

    def init(self):
        """Set ESC delay and then import curses."""
        global curses
        # Set ESC detection time
        os.environ["ESCDELAY"] = str(self.app.config["app"]["escdelay"])
        # Now import curses, otherwise ESCDELAY won't have any effect
        import curses
        import curses.textpad

    def run(self, func):
        """Run the application main function via the curses wrapper for safety."""
        curses.wrapper(func)

    def load(self, *args):
        """Setup curses."""
        # Log the terminal type
        self.app.log("Loading UI for terminal: " + curses.termname().decode("utf-8"), LOG_INFO)

        self.screen = curses.initscr()
        self.setup_colors()

        curses.cbreak()
        curses.noecho()
        try:
            # Might fail on vt100 terminal emulators
            curses.curs_set(0)
        except:
            self.app.log("curses.curs_set(0) failed!", LOG_WARNING)

        self.screen.keypad(1)

        self.current_yx = self.screen.getmaxyx()  # For checking resize
        self.setup_mouse()
        self.setup_windows()

    def unload(self):
        """Unload curses."""
        curses.endwin()

    def setup_mouse(self):
        # Mouse support
        curses.mouseinterval(10)
        if self.app.config["editor"]["use_mouse"]:
            curses.mousemask(-1)  # All events
        else:
            curses.mousemask(0)  # All events

    def setup_colors(self):
        """Initialize color support and define colors."""
        curses.start_color()
        try:
            curses.use_default_colors()
        except:
            self.app.logger.log("Failed to load curses default colors. You could try 'export TERM=xterm-256color'.")
            return False

        # Default foreground color (could also be set to curses.COLOR_WHITE)
        fg = -1
        # Default background color (could also be set to curses.COLOR_BLACK)
        bg = -1

        # This gets colors working in TTY's as well as terminal emulators
        # curses.init_pair(10, -1, -1) # Default (white on black)
        # Colors for xterm (not xterm-256color)
        # Dark Colors
        curses.init_pair(0, curses.COLOR_BLACK, bg)      # 0 Black
        curses.init_pair(1, curses.COLOR_RED, bg)        # 1 Red
        curses.init_pair(2, curses.COLOR_GREEN, bg)      # 2 Green
        curses.init_pair(3, curses.COLOR_YELLOW, bg)     # 3 Yellow
        curses.init_pair(4, curses.COLOR_BLUE, bg)       # 4 Blue
        curses.init_pair(5, curses.COLOR_MAGENTA, bg)    # 5 Magenta
        curses.init_pair(6, curses.COLOR_CYAN, bg)       # 6 Cyan
        curses.init_pair(7, fg, bg)                      # White on Black
        curses.init_pair(8, fg, curses.COLOR_BLACK)  # White on Black (Line number color)

        # Nicer shades of same colors (if supported)
        if curses.can_change_color():
            try:
                # TODO: Define RGB for these to avoid getting
                # different results in different terminals
                # xterm-256color chart http://www.calmar.ws/vim/256-xterm-24bit-rgb-color-chart.html
                curses.init_pair(0, 242, bg)  # 0 Black
                curses.init_pair(1, 204, bg)  # 1 Red
                curses.init_pair(2, 119, bg)  # 2 Green
                curses.init_pair(3, 221, bg)  # 3 Yellow
                curses.init_pair(4, 69, bg)   # 4 Blue
                curses.init_pair(5, 171, bg)  # 5 Magenta
                curses.init_pair(6, 81, bg)   # 6 Cyan
                curses.init_pair(7, 15, bg)   # 7 White
                # curses.init_pair(8, 8, bg)   # 8 Gray (Line number color)
                curses.init_pair(8, 8, curses.COLOR_BLACK)  # 8 Gray on Black (Line number color)
            except:
                self.app.logger.log("Enhanced colors failed to load. You could try 'export TERM=xterm-256color'.")
                self.app.config["editor"]["show_highlighting"] = False
        else:
            self.app.logger.log("Enhanced colors not supported. You could try 'export TERM=xterm-256color'.", LOG_INFO)
            self.app.config["editor"]["show_highlighting"] = False

        self.app.themes.use(self.app.config["editor"]["theme"])

    def setup_windows(self, resize=False):
        """Initialize windows."""
        yx = self.screen.getmaxyx()
        self.text_input = None
        self.header_win = curses.newwin(1, yx[1], 0, 0)
        self.status_win = curses.newwin(1, yx[1], yx[0]-1, 0)

        # Test for new curses
        if "get_wch" not in dir(self.header_win):
            # Notify only once
            if not self.warned_old_curses:
                self.app.log("Using old curses! Some keys and special characters might not work.", LOG_WARNING)
                self.warned_old_curses = 1

        y_sub = 0
        y_start = 0
        if self.app.config["display"]["show_top_bar"]:
            y_sub += 1
            y_start = 1
        if self.app.config["display"]["show_bottom_bar"]:
            y_sub += 1
        if self.app.config["display"]["show_legend"]:
            y_sub += 2
        self.editor_win = curses.newwin(yx[0]-y_sub, yx[1], y_start, 0)
        if self.app.config["display"]["show_top_bar"]:
            self.legend_win = curses.newwin(2, yx[1], yx[0]-y_sub+1, 0)
        else:
            self.legend_win = curses.newwin(2, yx[1], yx[0]-y_sub, 0)

        if resize:
            self.app.get_editor().resize((yx[0]-y_sub, yx[1]))
            self.app.get_editor().move_win((y_start, 0))

    def get_size(self):
        """Get terminal size."""
        y, x = self.screen.getmaxyx()
        return (x, y)

    def update(self):
        self.check_resize()

    def refresh(self):
        self.refresh_status()
        self.screen.refresh()

    def resize(self, yx=None):
        """Resize UI to yx."""
        if yx is None:
            yx = self.screen.getmaxyx()
        self.screen.erase()
        curses.resizeterm(yx[0], yx[1])
        self.setup_windows(resize=True)
        self.screen.refresh()

    def check_resize(self):
        """Check if terminal has resized."""
        yx = self.screen.getmaxyx()
        if self.current_yx != yx:
            self.current_yx = yx
            self.resize(yx)

    def refresh_status(self):
        """Refresh status windows."""
        if self.app.config["display"]["show_top_bar"]:
            self.show_top_status()
        if self.app.config["display"]["show_legend"]:
            self.show_legend()
        if self.app.config["display"]["show_bottom_bar"]:
            self.show_bottom_status()

    def show_top_status(self):
        """Show top status row."""
        self.header_win.erase()
        size = self.get_size()
        display = self.app.config["display"]
        head_parts = []
        if display["show_app_name"]:
            name_str = "Suplemon Editor v" + self.app.version + " -"
            if self.app.config["app"]["use_unicode_symbols"]:
                logo = "\u2688"      # Simple lemon (filled)
                name_str = " " + logo + " " + name_str
            head_parts.append(name_str)

        # Add module statuses to the status bar
        for name in self.app.modules.modules.keys():
            module = self.app.modules.modules[name]
            if module.options["status"] == "top":
                status = module.get_status()
                if status:
                    head_parts.append(status)

        if display["show_file_list"]:
            head_parts.append(self.file_list_str())

        head = " ".join(head_parts)
        head = head + (" " * (self.screen.getmaxyx()[1]-len(head)-1))
        if len(head) >= size[0]:
            head = head[:size[0]-1]
        if self.app.config["display"]["invert_status_bars"]:
            self.header_win.addstr(0, 0, head, curses.color_pair(0) | curses.A_REVERSE)
        else:
            self.header_win.addstr(0, 0, head, curses.color_pair(0))
        self.header_win.refresh()

    def file_list_str(self):
        """Return rotated file list beginning at current file as a string."""
        curr_file_index = self.app.current_file_index()
        files = self.app.get_files()
        file_list = files[curr_file_index:] + files[:curr_file_index]
        str_list = []
        no_write_symbol = ["!", "\u2715"][self.app.config["app"]["use_unicode_symbols"]]
        is_changed_symbol = ["*", "\u2732"][self.app.config["app"]["use_unicode_symbols"]]
        for f in file_list:
            prepend = [no_write_symbol, ""][f.is_writable()]
            append = ["", is_changed_symbol][f.is_changed()]
            fname = prepend + f.name + append
            if not str_list:
                str_list.append("[" + fname + "]")
            else:
                str_list.append(fname)
        return " ".join(str_list)

    def show_bottom_status(self):
        """Show bottom status line."""
        editor = self.app.get_editor()
        size = self.get_size()
        cur = editor.get_cursor()
        data = "@ " + str(cur[0]) + "," + str(cur[1]) + " " + \
            "cur:" + str(len(editor.cursors)) + " " + \
            "buf:" + str(len(editor.get_buffer()))
        if self.app.config["app"]["debug"]:
            data += " cs:"+str(editor.current_state)+" hist:"+str(len(editor.history))  # Undo / Redo debug
        # if editor.last_find:
        #     find = editor.last_find
        #     if len(find) > 10:find = find[:10]+"..."
        #     data = "find:'"+find+"' " + data

        # Add module statuses to the status bar
        for name in self.app.modules.modules.keys():
            module = self.app.modules.modules[name]
            if module.options["status"] == "bottom":
                data += " " + module.get_status()

        self.status_win.erase()
        status = self.app.get_status()
        extra = size[0] - len(status+data) - 1
        line = status+(" "*extra)+data

        if len(line) >= size[0]:
            line = line[:size[0]-1]
        if self.app.config["display"]["invert_status_bars"]:
            self.status_win.addstr(0, 0, line, curses.color_pair(0) | curses.A_REVERSE)
        else:
            self.status_win.addstr(0, 0, line, curses.color_pair(0))

        self.status_win.refresh()

    def show_legend(self):
        """Show keyboard legend."""
        # TODO: get from key bindings
        self.legend_win.erase()
        keys = [
            ("^S", "Save"),
            ("F1", "Save as"),
            ("F2", "Reload"),
            ("F5", "Undo"),
            ("F6", "Redo"),
            ("^O", "Open"),
            ("^C", "Cut"),
            ("^V", "Paste"),
            ("^F", "Find"),
            ("^D", "Find next"),
            ("^A", "Find all"),
            ("^W", "Duplicate line"),
            ("ESC", "Single cursor"),
            ("^G", "Go to"),
            ("^E", "Run command"),
            ("F8", "Mouse mode"),
            ("^X", "Exit"),
        ]
        x = 0
        y = 0
        max_y = 1
        for key in keys:
            if x+len(" ".join(key)) >= self.get_size()[0]:
                x = 0
                y += 1
                if y > max_y:
                    break
            self.legend_win.addstr(y, x, key[0], curses.A_REVERSE)
            x += len(key[0])
            self.legend_win.addstr(y, x, " "+key[1])
            x += len(key[1])+2
        self.legend_win.refresh()

    def show_capture_status(self, s="", value=""):
        """Show status when capturing input."""
        self.status_win.erase()
        self.status_win.addstr(0, 0, s, curses.A_REVERSE)
        self.status_win.addstr(0, len(s), value)

    def _process_query_key(self, key):
        """Process keystrokes from the Textbox window."""
        # TODO: implement this to improve interacting in the input box
        if self.app.config["app"]["debug"]:
            self.app.log("Query key input:"+str(key), LOG_INFO)
        return key

    def _query(self, text, initial=""):
        """Ask for text input via the status bar."""
        self.show_capture_status(text, initial)
        self.text_input = curses.textpad.Textbox(self.status_win)
        try:
            out = self.text_input.edit(self._process_query_key)
        except:
            return False

        # If input begins with prompt, remove the prompt text
        if len(out) >= len(text):
            if out[:len(text)] == text:
                out = out[len(text):]
        if len(out) > 0 and out[-1] == " ":
            out = out[:-1]
        out = out.rstrip("\r\n")
        return out

    def query(self, text, initial=""):
        """Get a single line input string from the user."""
        result = self._query(text, initial)
        return result

    def query_bool(self, text, default=False):
        """Get a boolean from the user."""
        indicator = "[y/N]"
        initial = ""
        if default:
            indicator = "[Y/n]"
            initial = "y"

        result = self._query(text + " " + indicator, initial)
        if result in ["Y", "y"]:
            return True
        if result == "":
            return default
        return False

    def get_input(self):
        """Get an input event from keyboard or mouse. Returns an InputEvent instance or False."""
        event = InputEvent()  # Initialize new empty event
        char = False
        input_func = None
        if "get_wch" in dir(self.screen):
            # New Python 3.3 curses method for wide characters.
            input_func = self.screen.get_wch
        else:
            # Old Python fallback. No multibyte characters.
            input_func = self.screen.getch
        try:
            char = input_func()
        except KeyboardInterrupt:
            # Handle KeyboardInterrupt as Ctrl+C
            event.set_key_name("^C")
            return event
        except:
            # No input available
            return False

        if char:
            if self.is_mouse(char):
                state = self.get_mouse_state()
                if state:
                    event.parse_mouse_state(state)
                    return event
            else:
                event.parse_key_code(char)
                return event
        return False

    def is_mouse(self, key):
        """Check for mouse events"""
        return key == curses.KEY_MOUSE

    def get_mouse_state(self):
        """Get the mouse event data."""
        try:
            mouse_state = curses.getmouse()
        except:
            self.app.log(get_error_info())
            return False
        # Translate the coordinates to the editor coordinate system
        return self._translate_mouse_to_editor(mouse_state)

    def _translate_mouse_to_editor(self, state):
        """Translate the screen coordinates to position in the editor view."""
        editor = self.app.get_editor()
        x, y = (state[1], state[2])
        if self.app.config["display"]["show_top_bar"]:
            y -= 1
        x -= editor.line_offset()
        y += editor.y_scroll
        return (state[0], x, y, state[3], state[4])
